### ISSUE TITLE
dask.array.Array producing wrong results in local threaded mode

### ISSUE THREAD

**Issue by mueslo** [dask/dask#10217]
**Describe the issue**:
I am trying to dask-ify a somewhat complex memory-intensive numpy+scipy calculation which in theory is easily parallelizable into chunks (since it is basically rastering over the first two dimensions), I am very new to dask, just started using it the past Friday! However, at random things will fail. I have not yet been able to find a minimal reproducing example that I can share here, but I am opening this issue already nevertheless, maybe some of the symptoms can help someone point me in the right direction for tracking down the origin of this bug. I am calling it a bug without having found a cause since the code seems to work perfectly fine without dask and also works fine with dask if I skip this particular operation, and also works fine in the single-threaded scheduler. 

```python
>>> b
<<< dask.array<reshape, shape=(7520000, 3), dtype=float64, chunksize=(2801200, 3), chunktype=numpy.ndarray>
>>> b2 = ((b @ inv_matrix) % 1 ) @ matrix  # modulo in a different basis, i.e. matrix, inv_matrix 3x3 numpy arrays with matrix @ inv_matrix ~= np.eye(3)
>>> b2 
<<< dask.array<getitem, shape=(7520000, 3), dtype=float64, chunksize=(2801200, 3), chunktype=numpy.ndarray>
```
calculations related to b2 start to gives a different result each time
```python
>>> ((b2-b)**2).sum(axis=1).max().compute()
<<< 184.83619308212212
>>> ((b2-b)**2).sum(axis=1).max().compute()
<<< 1400.402682918915
>>> ((b2-b)**2).sum(axis=1).max().compute()
<<< 180.68333699762377
>>> ((b2-b)**2).sum(axis=1).max().compute()
<<< 1427.8745037645813

>>> ((b2)**2).sum(axis=1).max().compute()
<<< 9.46157028666362
>>> ((b2)**2).sum(axis=1).max().compute()
<<< 9.46157028666362
>>> ((b2)**2).sum(axis=1).max().compute()
<<< 9.46157028666362
>>> ((b2)**2).sum(axis=1).max().compute()
<<< 9.46157028666362
>>> ((b2)**2).sum(axis=1).max().compute()
<<< 470.2506360498207
```
using the single-threaded scheduler this is NOT the case!, and anything related to b seems to give the same result each time

computation visualisation of b2:

![b2](https://user-images.githubusercontent.com/847751/233981230-c2543a3b-0a67-48ba-ab9a-66191fc52336.svg)

```

**Minimal Complete Verifiable Example**:

```python
# none yet, sorry :(
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2023.4.0
- Python version: 3.11
- Operating System: Debian 12 (bookworm)
- Install method (conda, pip, source): pip

**Comment by phofl** [comment]
Hi, thanks for your report. I can't reproduce on arm Mac. Do you have access to another machine?

**Comment by mueslo** [comment]
Yes, I just tried it on a remote compute node (intel x86_64) and a different python version (3.9) and the same thing occurs. My local machine is AMD x86_64. Added those details.

**Comment by jrbourbeau** [comment]
cc @ravwojdyla who authored https://github.com/dask/dask/pull/7000

**Comment by ravwojdyla** [comment]
> The reported bug above does not occur below some sort of chunk size threshold, presumably because the hypothetical race condition does not occur if synchronization dominates. 

👋 @mueslo can you share some more on this please?

<hr>

I can reproduce this on Mac/M2 and Linux/Intel x86_64. #7000 `matmul` implementation follows the "tensordot trick" https://github.com/dask/dask/issues/6874#issuecomment-735944467, and turns out that this issue also affects `tensordot`:

<details>

<summary> Code w/ tensordot</summary>

```py
import numpy as np
import dask.array as da
m1 = np.random.random((3, 3))
m2 = np.asfortranarray(np.random.random((3,3)))
print(np.isfortran(m1), np.isfortran(m2))

x = da.random.random((400, 400, 47, 3), chunks=(100, 400, 47, 3))
b = x.reshape(-1, 3)
# NOTE: dot is implemented via tensordot 
b2 = np.dot(np.dot(b, m1), m2)

print('[BUG] MT / ST')
while True:
    try:
        mt, st = b2.max().compute(), b2.max().compute(scheduler='single-threaded')
        print(f"[{'X' if mt != st else ' '}] {mt} / {st}")
    except KeyboardInterrupt:
        break

```
</details>

I can't reproduce this issue using einsum (matrix multiplication). Based on this I suspect the issue is likely related to the "tensordot trick" used in both `tensordot` and `matmul` (maybe others, that I don't know about). Maybe some kind of optimization kicks in with certain expectation on the intermediate results(?)

**Comment by mueslo** [comment]
> > The reported bug above does not occur below some sort of chunk size threshold, presumably because the hypothetical race condition does not occur if synchronization dominates.
> 
> wave @mueslo can you share some more on this please?

Hi, sorry for the delay. What I mean by that is that if you e.g. decrease the chunk size from my examples above (e.g. chunks=(10, 400, 47, 3)), then the error will occur much less frequently. Make them small enough, and you basically cannot get the error to pop up. Since qualitatively exactly the same thing should be happening as for my failing example, I assumed there must be some race condition somewhere. But I have absolutely no prior experience with dask (I encountered and opened this issue within 24h of first trying it), so I have no idea how the multithreaded scheduling internals actually look like.

**Comment by ravwojdyla** [comment]
@mueslo thanks for the extra context! To summarize there seems to be some kind of bug that affects at least `tensordot` and `matmul` for that specific scenario, and it's likely (at least that's my theory) related to the "tensordot trick".

**Comment by mueslo** [comment]
I have another small update, the reshape operation is actually not important, but rather it seems to only occur only for `ndim == 2` itself, somehow.

Using 
```python
b = da.random.random((400*400*47, 3), chunks=(100*400*47, 3))
b2 = (b @ m1) @ m2
```
Also produces the bug, an equivalent 

```python
n = 1
b = da.random.random((400*(400//n)*47, n, 3), chunks=(100*(400//n)*47, n, 3))
b2 = (b @ m1) @ m2
# (+ optional reshape to (-1, 3) afterwards)
```
however, does not, for any `n` even when `n == 1`! So it seems to be related to some special path that is taken in that case.
